# Laboratorio di Analisi Numerica
## Lezione 1

Cecilia Pagliantini <cecilia.pagliantini@unipi.it>  
Leonardo Robol <leonardo.robol@unipi.it>  
02 ottobre 2026

> **Quantità di esercizi:** in questa dispensa ci sono *più esercizi* di quanti uno studente medio riesca a farne durante una lezione di laboratorio, specialmente tenendo conto anche degli esercizi facoltativi. Questo perché è pensata per “tenere impegnati” per tutta la lezione anche quegli studenti che già hanno un solido background di programmazione.
>
> Quindi fate gli esercizi che riuscite, partendo da quelli *non* segnati come facoltativi, e non preoccupatevi se non li finite tutti!

## Installare Julia e usare il notebook

Julia e Jupyter sono disponibili sui computer dell'aula H. Dopo aver effettuato il login sul captive portal, aprite un terminale nella cartella del laboratorio, lanciate `jupyter notebook` e aprite questo file scegliendo il kernel Julia. In alternativa, potete usare un qualunque sistema di Notebook online, come [Google Colab](https://colab.research.google.com/), e caricare il notebook.

Per chi utilizza il proprio PC, il primo passo è installare Julia dal [sito ufficiale](https://julialang.org/downloads/). Per usare Julia con Jupyter occorre poi installare il pacchetto `IJulia`: avviate Julia e digitate `using Pkg; Pkg.add("IJulia")`. Un'alternativa valida è usare [VS Code](https://code.visualstudio.com/) o l'alternativa free software [VSCodium](https://vscodium.com/), che permettono di visualizzare, editare, ed eseguire il notebook nello stesso editor. L'ambiente Jupyter deve comunque essere installato.

Un notebook è composto da celle di testo, come questa, e celle di codice. Per eseguire la cella selezionata premete **Shift+Invio**. Le definizioni e le variabili restano disponibili nelle celle successive: per ottenere sempre risultati riproducibili, eseguite le celle dall'alto verso il basso. Se necessario usate **Kernel → Restart Kernel and Run All Cells**. Salvando il notebook, testo e codice rimangono insieme nel file `.ipynb`.

## 1. Primo programma

La cella seguente è l'equivalente del prompt interattivo di Julia. Eseguiamola per mostrare il nostro primo messaggio.

In [ ]:
println("Hello, world")

> **Julia:** Julia è un linguaggio gratuito e Open Source, distribuito con licenza MIT. Può essere usato liberamente sia a casa sia in laboratorio. Il linguaggio e il suo ecosistema sono sviluppati pubblicamente; la documentazione ufficiale è disponibile su [docs.julialang.org](https://docs.julialang.org/).
>
> Julia non è un clone di MATLAB/Octave: pur condividendo molte idee della programmazione numerica, presenta differenze di sintassi e di comportamento. In questo notebook useremo direttamente la sintassi Julia.

## 2. Primi calcoli in virgola mobile

Julia usa il tipo `Float64` per i letterali in virgola mobile in doppia precisione (8 byte per ogni numero).

Queste tre funzioni definiscono i limiti fisici di precisione e rappresentazione della memoria del computer per quel tipo di dato:
*  **floatmax(Float64):** È il valore finito più grande che il computer può rappresentare
*   **floatmin(Float64):** È il più piccolo valore positivo normalizzato che il computer può rappresentare vicino allo zero senza perdere precisione
*  **eps(Float64) (Epsilon di macchina):** È la distanza minima tra il numero $1.0$ e il numero rappresentabile immediatamente successivo \
*(avendo come valore $2.22 \times 10^{-16}$ significa che indica che i numeri a 64 bit mantengono circa 15-17 cifre significative di precisione. Se avessi avuto ad esempio $\approx 1.92 \times 10^{-34}$ allora avrei avuto 33-35 cifre significative)*



In [ ]:
@show floatmin(Float64)
@show floatmax(Float64)
@show eps(Float64);

Julia come una calcolatrice:

In [ ]:
@show 1 + 1
@show 10^10
@show 1e10
@show (1e10)^2;

1 + 1 = 2
10 ^ 10 = 10000000000
1.0e10 = 1.0e10
1.0e10 ^ 2 = 1.0e20


In Julia i vari dati hanno un tipo preciso. `10^10` è un calcolo tra interi, mentre `1e10` è un numero `Float64`.

Se c'è un punto e virgola alla fine della cella, Julia esegue il calcolo ma il notebook non ne visualizza il risultato. Per visualizzare un dato intermedio, possiamo usare la macro `@show` come abbiamo fatto sopra.

In [ ]:
1 + 1;

Perdita di precisione in alcuni calcoli: Julia mostra già, quando serve, un numero di cifre sufficiente a identificare il valore `Float64`; non serve alcun comando particolare per regolare il numero di cifre mostrate.

In [ ]:
a = 1e10
b = 1e4
c = (a + b)^2
@show a b c
c - a^2 - 2*a*b - b^2

a = 1.0e10
b = 10000.0
c = 1.000002000001e20


7936.0

La perdita di precisione deriva principalmente da sottrazioni tra due numeri grandi e molto vicini (*errori di cancellazione*), ma anche da moltiplicazioni:

In [ ]:
a = 98.0
@show 1 - a*(1/a)
a = 97.0
@show 1 - a*(1/a);

1 - a * (1 / a) = 1.1102230246251565e-16
1 - a * (1 / a) = 0.0


*(il risultato della prima operazione dovrebbe essere 0)*

Quando succede? Controlliamo con un breve programma. In MATLAB avremmo creato uno *script*, cioè un file `.m`; in questo notebook possiamo invece inserire la sequenza di comandi direttamente in una cella.

Dentro un ciclo il notebook non visualizza automaticamente il valore di un'espressione isolata, perciò usiamo `println` per scrivere il valore di `k`.

In [ ]:
for k = 1:300
    a = k * (1 / k)
    if a != 1
        println("k = ", k)
    end
end

k = 49
k = 98
k = 103
k = 107
k = 161
k = 187
k = 196
k = 197
k = 206
k = 214
k = 237
k = 239
k = 249
k = 253


Dovremmo ottenere `k = 49, 98, 103, 107, 161, 187, 196, 197, 206, 214, 237, 239, 249, 253`. Possiamo controllare che per i valori stampati a schermo il valore calcolato di `k*(1/k)` è diverso da 1.

Se preferite lavorare anche con file sorgente, potete salvare il codice in un file `.jl` e caricarlo con `include("nomefile.jl")`. Nel notebook, però, è sufficiente eseguire la cella: non è necessario creare un file separato né impostare una particolare cartella di lavoro.

## 3. Funzioni e accumulatori

In [ ]:
function fact(n)
    # calcola il fattoriale di n
    # n dev'essere un intero
    f = 1.0
    # f fa da accumulatore: parte da 1,
    # a ogni passo, lo moltiplico per k
    for k = 1:n
        f = f * k
    end
    # ora f vale n!
    return f
end

println("4! = ", fact(4))

4! = 24.0


In un notebook la funzione non va salvata in un file separato: dopo aver eseguito la cella che la definisce, possiamo richiamarla da qualunque cella successiva. L'inizializzazione con `1.0` fa sì che l'accumulatore sia un `Float64`, come i numeri usati nel programma MATLAB originale.

In [ ]:
fact(10)

3.6288e6

### Esercizio 1

Scrivete una funzione `pow(x, n)` che calcoli $x^n$. Inserite la soluzione nella cella seguente ed eseguitela: la funzione verrà usata nel resto del notebook.

In [ ]:
# Scrivete qui la funzione pow(x, n)

function pow(x, n)

  f = 1.0 #contatore

  for k = 1:n
    f = f * x
  end

  return f
end

println("3^3 = ", pow(3,3))

3^3 = 27.0


## 4. Calcolo dell'esponenziale

$x^n=1+\frac{x}{1!}+\dots+\frac{x^k}{k!}$

In [ ]:
function myexp(x, n)
    # calcola exp(x) con la serie di Taylor troncata all'n-esimo termine
    a = 1.0  # accumulatore
    for k = 1:n
        a = a + pow(x, k) / fact(k)
    end
    return a
end

Qualche esperimento su quanti termini servono per approssimare bene: confrontiamo con la funzione `exp(x)` di Julia, che calcola l'esponenziale meglio di noi.

In [ ]:
function myexp(x, n)
    # calcola exp(x) con la serie di Taylor troncata all'n-esimo termine
    a = 1.0  # accumulatore
    for k = 1:n
        a = a + pow(x, k) / fact(k)
    end
    return a
end

@show myexp(1.0, 5)
@show myexp(10.0, 50)
@show exp(10.0);
@show myexp(-20.0, 500);

myexp(1.0, 5) = 2.7166666666666663
myexp(10.0, 50) = 22026.46579480671
exp(10.0) = 22026.465794806718
myexp(-20.0, 500) = NaN


Due problemi:

- **Inaccurato:** provate `myexp(-20.0, 500)`.
- **Lento:** con $n$ termini, il numero di operazioni da eseguire cresce come $n^2$ (perché?).

Risolviamo il secondo problema introducendo un altro accumulatore:

anziche fare $a =a+\frac{x^k}{k!}$ creo $\frac{x^k}{k!}$ facendo $t=t\cdot\frac{x}{k}$ con $k$ che cresce e poi $a=a+t$ per creare la somma di taylor

In [ ]:
function myexp2(x, n)
    # calcola e^x con Taylor troncato
    # ma usa solo O(n) operazioni
    t = 1.0  # accumulatore che contiene il termine generico della sommatoria
    a = 1.0  # accumulatore che contiene le somme parziali
    for k = 1:n
        t = t * x / k
        a = a + t
    end
    return a
end

@show myexp(-3,500);
@show myexp(3,500);

myexp(-3, 500) = 0.049787068367863986
myexp(3, 500) = 20.08553692318766


Ora va meglio:

In [ ]:
@show myexp(-20.0, 500)
@show myexp2(-20.0, 500);

Cosa succedeva?

In [ ]:
@show fact(500)
@show pow(-20.0, 500)
@show Inf / Inf;

Ci sono ancora problemi di accuratezza sui numeri negativi:

In [ ]:
myexp2(-30.0, 500)

Un esponenziale dovrebbe sempre essere positivo… Ci sono pesanti errori di cancellazione nella formula che abbiamo usato (perché?).

$$
\text{Errore relativo:}\quad \frac{\left|\operatorname{myexp2}(x)-e^x\right|}{e^x}.
$$

La soluzione: cambiare algoritmo e sceglierne uno che non porti a errori di cancellazione.

In [ ]:
@show exp(-30.0)
@show myexp2(-30.0, 500)
@show 1 / myexp2(30.0, 500);

### Esercizio 2

Scrivere una funzione `myexp3(x)` che controlla se $x$ è negativo o positivo e, a seconda del segno, calcola l'esponenziale come $1/e^{-x}$ oppure $e^x$ con la serie di Taylor troncata a $n=500$.

In [ ]:
# Scrivete qui la funzione myexp3(x)
function myexp3(x)
    t = 1.0  # accumulatore che contiene il termine generico della sommatoria
    a = 1.0  # accumulatore che contiene le somme parziali

    if x<0
      for k = 1:500
        t = t * (-x) / k
        a = a + t
      end
    return 1/a
    end

    for k = 1:500
        t = t * x / k
        a = a + t
    end

    return a
end

@show myexp3(-3);
@show myexp3(3);

myexp3(-3) = 0.04978706836786396
myexp3(3) = 20.08553692318766


### Esercizio 3

Scrivere una funzione `solve2(a, b, c)` che risolva l'equazione di secondo grado $ax^2+bx+c=0$ nel modo più stabile possibile. Testare l'implementazione su $x^2-(10^{10}+10^{-10})x+1=0$.

**Suggerimento:** ci sono al massimo due sottrazioni. Una è necessaria (perché?); l'altra no.

In [ ]:
# Scrivete qui la funzione solve2(a, b, c) e il test richiesto

function solve2(a, b, c)
  delta = sqrt((b^2)-(4*a*c))
  # evito la cancellazione numerica usando sign(b) per avere sempre -b-delta o b+delta
  x2 = (-b-sign(b)*delta)/(2*a)
  x1=c/(a*x2)

return x1,x2
end

@show solve2(1,-1,-6);
@show solve2(1,-2,1);

solve2(1, -1, -6) = (-2.0, 3.0)
solve2(1, -2, 1) = (1.0, 1.0)


### Esercizio 4 (facoltativo)

Come si può calcolare $\frac{1-\cos x}{x^2}$ in modo accurato per valori di $x$ piccoli? Testate il risultato su $x=10^{-10}$.

In [ ]:
# Scrivete qui la soluzione dell'esercizio 4

### Esercizio 5 (facoltativo)

Si consideri la funzione $f(x)=\log(1+x)/x$ nell'intervallo $[10^{-15},10^{-14}]$.

1. Si confrontino i valori ottenuti su punti dell'intervallo calcolando $f(x)$ o
   $$g(x)=\frac{\log(1+x)}{(1+x)-1}.$$
2. Si faccia l'analisi teorica dell'errore allo scopo di capire il motivo di questo comportamento.

In [ ]:
# Scrivete qui la soluzione dell'esercizio 5

### Esercizio 6 (facoltativo)

Si supponga di voler modificare il calcolo dell'esponenziale attraverso l'uso di una funzione razionale, invece che della serie di Taylor troncata. Si determinino due polinomi $p_k(z)$ e $q_k(z)$ di grado al più $k$ tali che

$$
e^z-\frac{p_k(z)}{q_k(z)}\sim\mathcal O(z^{2k+1}),\qquad q_k(0)=1.
$$

Si provi a sostituire questa approssimazione all'interno dell'algoritmo `myexp`.

**Suggerimento:** se $q_k(0)=1$, la condizione si può riscrivere in modo equivalente come $p_k(z)-e^zq_k(z)\sim\mathcal O(z^{2k+1})$, che fornisce delle equazioni lineari per i coefficienti di $p_k(z)$ e $q_k(z)$.

In [ ]:
# Scrivete qui la soluzione dell'esercizio 6